doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,908 @@
+     */
+    int (*read)(ieee802154_dev_t *dev, void *buf, size_t size,
+                ieee802154_rx_info_t *info);
+```
+
+### PHY Operations
+```c
+    /**
+     * @brief Perform Standalone Clear Channel Assessment
+     *
+     * This function performs blocking CCA to check if the channel is clear.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     *
+     * @return true if channel is clear.
+     * @return false if channel is busy.
+     */
+    bool (*cca)(ieee802154_dev_t *dev);
+
+    /**
+     * @brief Set the threshold for the Energy Detection (first mode of CCA)
+     *
+     * @pre the device is not sleeping
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] threshold the threshold in dBm.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_cca_threshold)(ieee802154_dev_t *dev, int8_t threshold);
+
+    /**
+     * @brief Set CCA mode
+     *
+     * All radios MUST at least implement the first CCA mode (ED Threshold).
+     *
+     * @pre the device is not sleeping
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] mode the CCA mode
+     *
+     * @return 0 on success
+     * @return -ENOTSUP if the mode is not supported
+     * @return negative errno on error
+     */
+    int (*set_cca_mode)(ieee802154_dev_t *dev, ieee802154_cca_mode_t mode);
+
+    /**
+     * @brief Set IEEE802.15.4 PHY configuration (channel, TX power)
+     *
+     * This function SHOULD NOT validate the PHY configuration unless
+     * it is specific to the device. The upper layer is responsible of all kind
+     * of validations.
+     *
+     * @pre the device is not sleeping
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] conf the PHY configuration
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*config_phy)(ieee802154_dev_t *dev, ieee802154_phy_conf_t *conf);
+
+    /**
+     * @brief Set the transceiver PHY state
+     *
+     * @pre the device is not sleeping
+     *
+     * @note the implementation MUST block until the state change occurs.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] state the new state
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_trx_state)(ieee802154_dev_t *dev, ieee802154_trx_state_t state);
+
+```
+
+### Device State Management
+```c
+    /**
+     * @brief Set the sleep state of the device (sleep or awake)
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] sleep whether the device should sleep or not.
+     *
+     * @post the transceiver state is @ref IEEE802154_TRX_STATE_TRX_OFF,
+     *       regardless of the value of @p sleep
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_sleep)(ieee802154_dev_t *dev, bool sleep);
```

Comment:
Is this actually needed? How about renaming the `start()` to `on()` and providing an `off()` function in addition?

To me, there is little to no different between starting a network device after `init()` or after it has been put into lowest power mode.

Reply:
Actually I like the `on/off` approach better than the `start`. I just wanted to avoid extra functions, but I think it would give even more control

Reply:
actually, what do you think of moving the "set_sleep" function to "off"?

Basically the behavior would be the same

Reply:
(I think that's what you meant, am I right?)

Reply:
note to self: since we are now using "on" and "off", the "on" function must initialize all values with the default "IEEE802154_XXX" params (e.g CONFIG_IEEE802154_DEFAULT_CHANNEL). Radios lose register content when turned off

Reply:
hmmm maybe it's even better to not set anything. There are some changes the MAC already will set a different PHY state than the default one when waking up the radio